### PR TITLE
[WIP] Add `StatefulDataLoader` to all recipes except knowledge_single

### PR DESCRIPTION
**Pull request by krammnic**
#### Context
What is the purpose of this PR? Is it to
- [X] add a new feature
- [ ] fix a bug
- [ ] update tests and/or documentation
- [ ] other (please add here)

Please link to any issues this PR addresses.
#2431  
#2439

#### Changelog
What are the changes made in this PR?
Added `StatefulDataLoader` to all recipes except `knowledge_single` (this recipe will be updated by @jxtngx)

Notice, that I haven't done verifying runs yet. Will add them gradually asap.

#### Test plan
Please make sure to do each of the following if applicable to your PR. If you're unsure about any one of these just ask and we will happily help. We also have a [contributing page](https://github.com/pytorch/torchtune/blob/main/CONTRIBUTING.md) for some guidance on contributing.

- [X] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [ ] add [unit tests](https://github.com/pytorch/torchtune/tree/main/tests/torchtune) for any new functionality
- [ ] update [docstrings](https://github.com/pytorch/torchtune/tree/main/docs/source) for any new or updated methods or classes
- [ ] run unit tests via `pytest tests`
- [ ] run recipe tests via `pytest tests -m integration_test`
- [ ] manually run any new or modified recipes with sufficient proof of correctness
- [ ] include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)

#### UX
If your function changed a public API, please add a dummy example of what the user experience will look like when calling it.
Here is a [docstring example](https://github.com/pytorch/torchtune/blob/6a7951f1cdd0b56a9746ef5935106989415f50e3/torchtune/modules/vision_transformer.py#L285)
and a [tutorial example](https://pytorch.org/torchtune/main/tutorials/qat_finetune.html#applying-qat-to-llama3-models)

- [ ] I did not change any public API
- [ ] I have added an example to docs or docstrings

### CODE REVIEW

File: recipes/lora_dpo_single_device.py [hunk 1]
```diff
@@ -245,10 +245,17 @@ def setup(self, cfg: DictConfig) -> None:
 
         # Dataloader depends on the tokenizer and loss_fn and should be
         # setup after all of these are setup
-        self._sampler, self._dataloader = self._setup_data(
+        collate_name = cfg.get("collate_fn", "torchtune.data.padded_collate_sft")
```
Comment (felipemello1) [inline recipes/lora_dpo_single_device.py:248]: this is not being passed to the setup_data. The default should also be "padded_collate_dpo".

Do you mind checking the collate_name in all recipes you changed? I think that this is an error in others too

File: recipes/qat_lora_finetune_distributed.py [hunk 2]
```diff
@@ -644,16 +647,16 @@ def _setup_data(
             raise RuntimeError("left_pad_sequence collator is only for inference.")
         collate_fn = _get_component_from_path(collate_fn)
 
-        sampler = DistributedSampler(
-            ds, num_replicas=self.world_size, rank=self.rank, shuffle=shuffle, seed=0
+        sampler = StatefulDistributedSampler(
```
Comment (felipemello1) [inline recipes/qat_lora_finetune_distributed.py:650]: we should also be adding "seed" to every sampler. I think that we also missed it in joe's original PR. Not a bug, since they hardcode seed=0, but it should probably use the same seed as the one given in the config.

File: recipes/qat_lora_finetune_distributed.py [hunk 3]
```diff
@@ -799,20 +804,10 @@ def train(self) -> None:
...
-
             pbar = tqdm(total=self._steps_per_epoch, disable=not (self.rank == 0))
-            for idx, batch in enumerate(self._dataloader):
-                if (
-                    self.max_steps_per_epoch is not None
-                    and (idx // self._gradient_accumulation_steps)
-                    == self.max_steps_per_epoch
-                ):
-                    break
+            self._dataloader.sampler.set_epoch(curr_epoch)
```
Comment (felipemello1) [inline recipes/qat_lora_finetune_distributed.py:808]: some recipes are missing this

File: tests/recipes/test_lora_dpo_distributed.py [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+        ckpt_path = Path(CKPT_MODEL_PATHS[ckpt])
+        ckpt_dir = ckpt_path.parent
+        log_file = gen_log_file_name(tmpdir)
+
+        # Config file needed for model conversion.
+        # Create a second copy for training resume
+        write_hf_ckpt_config(ckpt_dir)
+        write_hf_ckpt_config(tmpdir)
+
+        # Train for two epochs
```
Comment (felipemello1) [inline tests/recipes/test_lora_dpo_distributed.py:81]: I dont see neither epoch or max_steps_per_epoch in this config. Shouldnt we hardcode these and make max_steps_per_epoch something small?

File: tests/recipes/test_ppo_full_finetune_single_device.py [hunk 5]
```diff
@@ -123,18 +123,18 @@ def test_loss(self, tmpdir, monkeypatch):
 
         loss_values = get_loss_values_from_metric_logger(log_file)
         expected_loss_values = [
```
Comment (felipemello1) [inline tests/recipes/test_ppo_full_finetune_single_device.py:125]: how were those generated?

File: tests/recipes/test_qat_distributed.py [hunk 6]
```diff
@@ -45,8 +45,58 @@ def _get_test_config_overrides(self):
...
+                10.458215713500977,
+                10.477818489074707,
+                10.396238327026367,
+                10.40851879119873,
+                10.433064460754395,
+                10.500737190246582,
+                10.483240127563477,
+                10.43812084197998,
+            ],
+            "llama3": [
```
Comment (felipemello1) [inline tests/recipes/test_qat_distributed.py:74]: how did we generate these?

File: recipes/knowledge_distillation_distributed.py [hunk 7]
```diff
@@ -680,11 +687,15 @@ def _setup_data(
...
                 )
             ),
+            # dropping last avoids shape issues with compile + flex attention
+            drop_last=True,
         )
 
-        utils.log_rank_zero(log, "Dataset and Sampler are initialized.")
+        if dataloader_state_dict is not None:
+            dataloader.load_state_dict(dataloader_state_dict)
+            list(dataloader)  # Hack to force dataloader to finish iteration
```
Comment (felipemello1) [inline recipes/knowledge_distillation_distributed.py:696]: if tests pass here, we have to delete this section in every recipe: https://github.com/pytorch/torchtune/pull/2490

File: tests/recipes/test_ppo_full_finetune_single_device.py (hunk 5 above)
Comment (SalmanMohammadi) [reply tests/recipes/test_ppo_full_finetune_single_device.py:125]: @krammnic how come these values were changed?

Comment (krammnic) [reply tests/recipes/test_ppo_full_finetune_single_device.py:125]: @SalmanMohammadi Maybe I misunderstood the question, but it was just a run of integration test for ppo

**Review by joecummings — APPROVED** [review]
